cryptography/src/commonMain/kotlin/com/wire/kalium/cryptography/utils/ChaCha20Utils.kt

```diff
@@ -0,0 +1,164 @@
+import com.ionspin.kotlin.crypto.secretstream.crypto_secretstream_xchacha20poly1305_TAG_MESSAGE
+import com.wire.kalium.cryptography.backup.Backup
+import com.wire.kalium.cryptography.backup.Backup.Companion.BACKUP_FILE_HEADER_LENGTH
+import com.wire.kalium.cryptography.kaliumLogger
+import okio.Buffer
+import okio.Sink
+import okio.Source
+import okio.buffer
+import kotlin.random.Random
+import kotlin.random.nextUBytes
+
+@OptIn(ExperimentalUnsignedTypes::class)
+internal class ChaCha20Utils {
+
+    @Suppress("TooGenericExceptionCaught")
+    suspend fun encryptBackupFile(
+        backupDataSource: Source,
+        outputSink: Sink,
+        backup: Backup
+    ): Long {
+
+        initializeLibsodiumIfNeeded()
+        var encryptedDataSize = 0L
+        val additionalInformation: UByteArray = ubyteArrayOf()
+
+        try {
+            val chaCha20Key = backup.generateChaCha20Key()
+            val hashedUserId = backup.hashUserId()
+            val backupHeader = backup.provideHeaderBuffer(hashedUserId).also {
+                if (it.isEmpty()) throw IllegalStateException("Backup header is empty")
+            }
+
+            // We append all the metadata unencrypted to the beginning of the backup file data
+            val outputBuffer = outputSink.buffer()
+            outputBuffer.write(backupHeader)
+            outputBuffer.flush()
+
+            val stateAndHeader = SecretStream.xChaCha20Poly1305InitPush(chaCha20Key)
+            val state = stateAndHeader.state
+            val chachaHeader = stateAndHeader.header.toByteArray()
+
+            // We write the ChaCha20 generated header prior to the encrypted backup file data
+            outputBuffer.write(chachaHeader)
+            outputBuffer.flush()
+
+            val inputContentBuffer = Buffer()
+            var byteCount: Long
+            while (backupDataSource.read(inputContentBuffer, ENCRYPT_BUFFER_SIZE).also { byteCount = it } != -1L) {
+                // We need to inform the end of the encryption stream with the TAG_FINAL
+                val appendingTag = if (byteCount < ENCRYPT_BUFFER_SIZE) { // TODO: Find a better way to detect the end of the stream
+                    crypto_secretstream_xchacha20poly1305_TAG_FINAL
+                } else
+                    crypto_secretstream_xchacha20poly1305_TAG_MESSAGE
+
+                val dataToEncrypt = inputContentBuffer.readByteArray(byteCount)
+                val uByteDataToEncrypt = dataToEncrypt.toUByteArray()
+
+                val encryptedData = SecretStream.xChaCha20Poly1305Push(
+                    state,
+                    uByteDataToEncrypt,
+                    additionalInformation,
+                    appendingTag.toUByte()
+                )
+                val tempWriteBuffer = Buffer()
+                tempWriteBuffer.write(encryptedData.toByteArray())
+                outputSink.write(tempWriteBuffer, tempWriteBuffer.size)
```

Comment:
Why do we need this tempWriteBuffer?

Reply:
Just to have a clean buffer to use for writing the recently encoded data.

Reply:
What else do you suggest to use instead?

Reply:
You can write the `encryptedData` directly to the outputSink if it's a `BufferedSink`. I see that that it isn't but it probably should be.

Reply:
I could reuse the `outputBuffer` and call it like this:
```
  outputBuffer.write(encryptedData.toByteArray())
  outputSink.write(outputBuffer.buffer, chunkWriteBuffer.size)
```
The problem with that, is that `outputBuffer` has a fixed allocated size determined by Okio under the hood (8 kilobytes if I'm not mistaken), so atm this could work. However, this will not guarantee that in the future they decide to lower it down in the future. Causing that way some issues in our implementation if this default size is 4k or lower, as we also use an `ENCRYPT_BUFFER_SIZE` of 4k, which will be insufficient for the first iteration of the loop where we need a size of `ENCRYPT_BUFFER_SIZE` + `BackupHeader.size` + `ChaCha20Header.size`.

That's why I prefer to have a dedicated buffer only for this type of write action.

Reply:
We don't need any buffers if you have a BufferedSink and there's no limits. See https://square.github.io/okio/3.x/okio/okio/okio/-buffered-sink/index.html